### PR TITLE
Add presigned URLs to erlcloud

### PR DESCRIPTION
**Pull request by lbakerchef**
This PR is part of a broader effort to replace AWS sigv2 with sigv4, encompassing modifications to oc_erchef, bookshelf, mini_s3, erlcloud, pedant clients, and other systems. Erlcloud received significant modifications for presigning of URLs, date and time mocking/modification, injection of host headers and ports, presigning with a' la carte headers, and expiration windows, among others. New unit tests exercising various aspects of newly-added functionality were also added.

Should this be merged, or (more likely) **left as its own branch**?  Unless someone believes otherwise, I'm recommending leaving this as a separate branch.  

See:
https://github.com/chef/chef-server/issues/1911
https://github.com/chef/chef-server/pull/1993
chef/mini_s3#31

- [x] Green verify pipeline
- [x] Green omnibus/adhoc pipeline
- [x] Green integration test pipeline
- [x] Green dialyzer pass

### CODE REVIEW

File: src/erlcloud_aws.erl [hunk 1]
```diff
@@ -989,22 +996,41 @@ sign_v4_headers(Config, Headers, Payload, Region, Service) ->
...
-                   Token -> [{"x-amz-security-token", Token} | Headers1]
+    sign_v4(Method, Uri, Config, Headers, Payload, Region, Service, QueryParams, Date).
+
+-spec sign_v4(atom(), list(), aws_config(), headers(), string() | binary(), string(), string(), list(), string()) -> headers().
+sign_v4(Method, Uri, Config, Headers0, Payload, Region, Service, QueryParams, Date0) ->
+    % use passed-in x-amz-date header or create one
+    Headers1 =
+        case proplists:get_value("x-amz-date", Headers0) of
+            undefined ->
+                Date = Date0,
```
Comment (stevendanna) [inline src/erlcloud_aws.erl]: Is this unused?

File: src/erlcloud_aws.erl [hunk 2]
```diff
@@ -989,22 +996,41 @@ sign_v4_headers(Config, Headers, Payload, Region, Service) ->
...
-    {PayloadHash, Headers1} =
-        sign_v4_content_sha256_header( [{"x-amz-date", Date} | Headers], Payload ),
-    Headers2 = case Config#aws_config.security_token of
-                   undefined -> Headers1;
-                   Token -> [{"x-amz-security-token", Token} | Headers1]
+    sign_v4(Method, Uri, Config, Headers, Payload, Region, Service, QueryParams, Date).
+
+-spec sign_v4(atom(), list(), aws_config(), headers(), string() | binary(), string(), string(), list(), string()) -> headers().
+sign_v4(Method, Uri, Config, Headers0, Payload, Region, Service, QueryParams, Date0) ->
+    % use passed-in x-amz-date header or create one
```
Comment (stevendanna) [inline src/erlcloud_aws.erl:1153]: I find the semantics of this a _little_ strange in that we are passing an explicit date but the x-amz-date command overrides this. I assume this is because of the double-duty this is doing for verifying our signatures (which will likely have the date header coming in via Headers) and generating new signatures (which I am assuming won't have the date header coming in via Headers)

File: src/erlcloud_s3.erl [hunk 3]
```diff
@@ -1084,17 +1088,95 @@ make_link(Expire_time, BucketName, Key, Config) ->
...
+
+-spec signature(aws_config(), string(), string(), string(), atom(), proplist(), proplist(), string()) -> string().
+signature(Config, Path, Date, Region, Method, QueryParams, Headers, Payload) ->
+  Service = "s3",
+  CredentialScope = erlcloud_aws:credential_scope(Date, Region, Service),
+  {CanonicalRequest, SignedHeaders} = erlcloud_aws:canonical_request(Method, Path, QueryParams, Headers, Payload),
+  ToSign = erlcloud_aws:to_sign(Date, CredentialScope, CanonicalRequest),
+  SigningKey = erlcloud_aws:signing_key(Config, Date, Region, Service),
+  [Result] = erlcloud_aws:base16(erlcloud_util:sha256_mac(SigningKey, ToSign)),
+  Result.
```
Comment (stevendanna) [inline src/erlcloud_s3.erl:1145]: I don't see many tests for this function. It might be nice to generate some pre-signed URLs with another tool and then use the results of that as test data that we could use to validate this function.

**Review by stevendanna — COMMENTED** [review]
Overall this one looks good to me.  I'd like to see a few more tests and I think we might need to clean up a bit how we handle the Host headers; but I'd like to see the call sites before making that call.

File: src/erlcloud_aws.erl (hunk 2 above)
Comment (lbakerchef) [reply src/erlcloud_aws.erl:1153]: I don't know if this helps, but the way it works is: you pass in a date, but if there is a date header also being passed in, the date in the date header is used.  If not, the passed-in date is used.

File: src/erlcloud_aws.erl (hunk 1 above)
Comment (lbakerchef) [reply src/erlcloud_aws.erl]: Date? Headers1? They are used.

File: src/erlcloud_s3.erl (hunk 3 above)
Comment (lbakerchef) [reply src/erlcloud_s3.erl:1145]: I don't think this is my code.  Would you still like me to add tests for it?

Comment (PrajaktaPurohit) [reply src/erlcloud_s3.erl:1145]: yeah - that would be really nice if we can add tests, that might also make it easier to get the commit merged upstream.

File: rebar.config [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 {erl_opts,
  [debug_info,
-  {platform_define, "^19.|^2", 'ERLANG_OTP_VERSION_19'},
+%  {platform_define, "^19.|^2", 'ERLANG_OTP_VERSION_19'},
```
Comment (PrajaktaPurohit) [inline rebar.config]: We are currently on 22.

Comment (lbakerchef) [reply rebar.config]: Can we just remove the comment, or do you want something like:
`{platform_define, "^22", 'ERLANG_OTP_VERSION_22'},`

File: src/erlcloud_aws.erl (hunk 2 above)
Comment (lbakerchef) [reply src/erlcloud_aws.erl:1153]: I think your 'double-duty' comment is probably correct.

File: src/erlcloud_s3.erl (hunk 3 above)
Comment (lbakerchef) [reply src/erlcloud_s3.erl:1145]: "...generate some pre-signed URLs with another tool..."

1) Does anyone know of a trusted tool that does this?

2) In theory we could simply take presigned URLs that worked against S3 in the past (meaning, we know they are valid) and run them through again, but the problem is, they expire.  The date looks mockable so that might not be an issue.  I'd have to create a build that spits out the inputs+outputs that came into/out-of this function which were then subsequently used to hit S3 with presigned URLs successfully.

**Review by newblaque — APPROVED** [review]

File: src/erlcloud_s3.erl (hunk 3 above)
Comment (lbakerchef) [reply src/erlcloud_s3.erl:1145]: Ok, I did #2 above.  I created a build to log diagnostics, and ran `chef-server-ctl test` against S3.  I was able to get the inputs and outputs to the signature function to create tests.  The tests worked.  However, my S3 `access_key_id` is in the source code.  Perhaps a better approach is to run against Bookshelf instead, although that is using our own code to verify our own code.

File: rebar.config (hunk 4 above)
Comment (lbakerchef) [reply rebar.config]: Also, if we plan to try to merge into their library in the future, do we want to change this?  They may want to support more platforms?

File: src/erlcloud_s3.erl (hunk 3 above)
Comment (lbakerchef) [reply src/erlcloud_s3.erl:1145]: Rewrote to use Bookshelf vs S3.

File: rebar.config (hunk 4 above)
Comment (lbakerchef) [reply rebar.config]: Done.

Comment (lbakerchef) [reply rebar.config]: I think the line restricts platforms to OTP 19.  I removed it so that it will run on whatever platform.  I tested it on 22 and 24.

**Review by vinay-satish — APPROVED** [review]
Code looks good to me, could you please add the links to the build kite pipelines.

**Comment by lbakerchef** [comment]
> Code looks good to me, could you please add the links to the build kite pipelines.

@vinay-satish See this PR for adhocs and umbrellas:
https://github.com/chef/chef-server/pull/2097

File: src/erlcloud_s3.erl [hunk 5]
```diff
@@ -506,7 +511,7 @@ put_bucket_lifecycle(BucketName, Policy, Config)
     put_bucket_lifecycle(BucketName, list_to_binary(XmlPolicy), Config);
 put_bucket_lifecycle(BucketName, XmlPolicy, Config)
   when is_list(BucketName), is_binary(XmlPolicy), is_record(Config, aws_config) ->
-    Md5 = base64:encode(crypto:hash(md5, XmlPolicy)),
+    Md5 = base64:encode(erlang:md5(XmlPolicy)),
```
Comment (marcparadise) [inline src/erlcloud_s3.erl:514]: Why the change from `crypto:hash` to `erlang:md5`? I distantly remember we had to go in the other direction once, years ago.

Comment (marcparadise) [inline src/erlcloud_s3.erl:514]: Why the change from crypto:hash to erlang:md5? I distantly remember we had to go in the other direction once, years ago.

**Review by marcparadise — APPROVED** [review]

File: src/erlcloud_s3.erl (hunk 5 above)
Comment (lbakerchef) [reply src/erlcloud_s3.erl:514]: The commit says that it is to fix a FIPS issue on rhel boxes.  If i recall correctly, @PrajaktaPurohit may know the context about this and know which one is correct - the `crypto:hash` or `erlang:md5`.
